IPIP 0001: Lightweight "RFC" Process for IPFS Specifications (#289)
ipfs/specs: RFC/0001-lightweight-rfc-process.md

# RFC 0001: Lightweight RFC Process for IPFS Specifications

- Start Date: 2022-06-10
- Related Issues:
  - [ipfs/specs/issues/286](https://github.com/ipfs/specs/issues/286)

# Summary

> This Request for Change (RFC) introduces lightweight RFC process

Reviewer: ```suggestion
This Request for Change (RFC) introduces a lightweight RFC process
```

> for IPFS specifications [repository][1].

Reviewer: ```suggestion
for the IPFS specifications [repository][1].
```

[1]: https://github.com/ipfs/specs/

# Motivation

> Protocol design discussions often take place in a repository of an IPFS

Reviewer: ```suggestion
Today, protocol design discussions often take place in a repository of an IPFS
```

> implementation. Often lacking useful input from IPFS Stewards, other

Reviewer: ```suggestion
implementation. These conversations are unintentionally obscured from the useful input of IPFS Stewards, other
```

implementations, service operators and the wider IPFS Community.

> IPFS Project needs a mechanism for proposing and evaluating specification

Reviewer: ```suggestion
The IPFS Project needs a mechanism for proposing and evaluating specification
```

improvements that are not tied to a specific programming language
or implementation of IPFS.

# Detailed design

> Adopt an informal RFC process for [ipfs/specs][1] repository, providing

Reviewer: ```suggestion
Adopt an informal RFC process for the [ipfs/specs][1] repository, providing a
```

minimal structure for opening, reviewing, and merging specification changes.

## Opening

Changes to IPFS specifications can be proposed by opening a PR against the
repository, and including a short "request for change" document based on the
template in `RFC/0000-template.md`.

## Reviewing

> IPFS Stewards will review new RFC PRs during weekly triage.

Reviewer: Do you want to explicitly list the github handles of the Stewards? Or punt until later.

Author: Punt. This is one of the topics I hope to flesh during our meetup in July and beyond, after we get some experience with RFC process.

For now, we will be reusing the existing [`@ipfs/w3dt-stewards`](https://github.com/orgs/ipfs/teams/w3dt-stewards/members)  team.

Author: Good idea. Created  @ipfs/specs-stewards  – figuring out who should stay/be added is discussed in https://github.com/ipfs/specs/issues/291

IPFS Community is encouraged to participate in the review process.

> Final decision belongs to IPFS Stewards.

Reviewer 2: ```suggestion
Final decision belongs to [IPFS Stewards].
```

## Merging

> RFC can be merged only after two Stewards approve it.

Reviewer 3: - What does it mean to be merged?
- What does it mean if an RFC is closed?

Author: I clarified we mean a git pull-request with changes that include an improvement proposal document (IPIP).

> RFC can be merged only after two Stewards approve it.

Reviewer 2: ```suggestion
RFC can be merged only after two [IPFS Stewards] approve it.
```

RFC number is assigned before the merge.

> RFC author and two approving Stewards are added to CODEOWNERS file to be

Reviewer 2: ```suggestion
RFC author and two approving [IPFS Stewards] are added to CODEOWNERS file to be
```

automatically asked to review any future changes to files added or modified
by the RFC.

## Long-term plan

> IPFS Stewards will adjust the process based on usage patterns.

Reviewer 2: ```suggestion
[IPFS Stewards] will adjust the process based on usage patterns.
```

## Design rationale

> We want to empower IPFS community and implementers with the ability to propose

Reviewer: ```suggestion
We want to empower IPFS community members and implementers with the ability to propose
```

changes in a well-known way, without introducing much overhead.

Guiding principles:
- No new tooling
  - Reuse Markdown, Git, and existing PR review process
- Convention over Byzantine process
  - Proposing a new RFC should have low cognitive overhead, allowing us to
    focus on specs

>   - Reuse existing Git developer accounts and reputation attached to them

Reviewer: ```suggestion
  - Reuse existing Github developer accounts and reputation attached to them
```

  - One should be able to create a valid RFC without reading a long explainer
    like this one. Looking at past RFCs, and then copying a template and

>     opening a PR with their proposal should be more than enough.

Reviewer: ```suggestion
    opening a PR with the proposal should be more than enough.
```

### User benefit

End users will indirectly benefit from a healthy RFC process being in place:

- IPFS community members will be able to use RFC drafts for evaluating ideas
  before investing time into building new things.
- The bar for creating a brand new IPFS implementation will be lowered, and
  existing implementations will be able to propose improvements for others to
  adopt. This removes the soft vendor lock-in present when the oldest
  implementation is considered as the reference standard and source of truth.

> - IPFS Implementers will have a better understanding why certain design

Reviewer: ```suggestion
- IPFS implementers will have a better understanding of why certain design
```
lowercase the "i" here since it's not a formal group name

  decisions were made, and have both historical context and language-agnostic
  specifications with test fixtures ready for use in their project, ensuring
  a high level of interoperability.
- More eyes looking at specifications will improve overall quality over time.

As a result, IPFS will become easier to implement, useful in more contexts,

> benefiting more people.

Reviewer: ```suggestion
and benefit more people.
```

### Compatibility

Existing contents of [ipfs/specs][1] repository act as the initial state
against which RFC PRs can be opened.

### Security

Existing Git-based review infrastructure, user accounts and reputation
system will be reused.

> Merging RFC will require approval from two IPFS Stewards.

Reviewer 2: ```suggestion
Merging RFC will require approval from two [IPFS Stewards].
```

### Alternatives

- Maintaining the status quo (no RFC process) is not acceptable, as we want to
  move specification discussions away from repositories of specific
  implementations. We need a mechanism for discussing improvements that is not
  tied to specific implementation or language.
- Creating more elaborate RFC process. This comes with increased overhead and
  risk. Introducing a complex process requires deeper understanding of
  community needs and pitfalls of preexisting processes, and since we don't
  have any process in place, starting light, limits the risk.

### Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

[IPFS Stewards]: https://linkToIssueWhereThisWillBeDetermined
```
